File: backend/app/api/routes/comments.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.schemas.comment import (
    CommentCreate,
    CommentResponse,
)
from app.services.comment import (
    create_comment,
    get_task_comments,
)

# Use the same authentication dependency
# that your existing task routes use.
from app.api.dependencies import get_current_user_id

from app.services.task import get_task


router = APIRouter(
    prefix="/api/v1/workspaces",
    tags=["Comments"],
)


@router.post(
    "/{workspace_id}/projects/{project_id}/tasks/{task_id}/comments",
    response_model=CommentResponse,
)
def add_comment(
    workspace_id: UUID,
    project_id: UUID,
    task_id: UUID,
    data: CommentCreate,
    db: Session = Depends(get_db),
    current_user_id: str = Depends(
        get_current_user_id
    ),
):
    task = get_task(
        db=db,
        workspace_id=str(workspace_id),
        project_id=str(project_id),
        task_id=str(task_id),
    )

    if not task:
        raise HTTPException(
            status_code=404,
            detail="Task not found",
        )

    if not data.content.strip():
        raise HTTPException(
            status_code=400,
            detail="Comment cannot be empty",
        )

    return create_comment(
        db=db,
        task=task,
        user_id=current_user_id,
        content=data.content.strip(),
    )


@router.get(
    "/{workspace_id}/projects/{project_id}/tasks/{task_id}/comments",
    response_model=list[CommentResponse],
)
def list_comments(
    workspace_id: UUID,
    project_id: UUID,
    task_id: UUID,
    db: Session = Depends(get_db),
    current_user_id: str = Depends(
        get_current_user_id
    ),
):
    task = get_task(
        db=db,
        workspace_id=str(workspace_id),
        project_id=str(project_id),
        task_id=str(task_id),
    )

    if not task:
        raise HTTPException(
            status_code=404,
            detail="Task not found",
        )

    return get_task_comments(
        db=db,
        task_id=str(task_id),
        workspace_id=str(workspace_id),
    )

File: frontend/src/services/comments.ts

import { api } from "./api";

/* ============================================================
   TYPES
============================================================ */

export interface TaskComment {
  id: string;
  task_id: string;
  workspace_id: string;
  user_id: string;
  content: string;
  created_at: string;
  updated_at: string;
}

export interface TaskActivity {
  id: string;
  task_id: string;
  workspace_id: string;
  user_id: string;
  action: string;
  details?: Record<string, unknown> | null;
  created_at: string;
}

/* ============================================================
   GET COMMENTS
============================================================ */

export async function getTaskComments(
  workspaceId: string,
  projectId: string,
  taskId: string,
): Promise<TaskComment[]> {
  const response = await api.get<TaskComment[]>(
    `/api/v1/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/comments`,
  );
  return response.data;
}

/* ============================================================
   ADD COMMENT
============================================================ */

export async function addTaskComment(
  workspaceId: string,
  projectId: string,
  taskId: string,
  content: string,
): Promise<TaskComment> {
  const response = await api.post<TaskComment>(
    `/api/v1/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/comments`,
    { content },
  );
  return response.data;
}

/* ============================================================
   GET TASK ACTIVITY LOGS
============================================================ */

export async function getTaskActivities(
  workspaceId: string,
  projectId: string,
  taskId: string,
): Promise<TaskActivity[]> {
  const response = await api.get<TaskActivity[]>(
    `/api/v1/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/activity`,
  );
  return response.data;
}
